be2-scala/src/test/scala/ch/epfl/pop/json/HighLevelProtocolSuite.scala

```diff
@@ -3,30 +3,122 @@ package ch.epfl.pop.json
+                               |      "witness_signatures": []
+                               |    }""".stripMargin
 
-    val expected: Message = Message(
-      Base64Data("eyJjcmVhdGlvbiI6MTYzMTg4NzQ5NiwiaWQiOiJ4aWdzV0ZlUG1veGxkd2txMUt1b0wzT1ZhODl4amdYalRPZEJnSldjR1drPSIsIm5hbWUiOiJoZ2dnZ2dnIiwib3JnYW5pemVyIjoidG9fa2xaTHRpSFY0NDZGdjk4T0xOZE5taS1FUDVPYVR0YkJrb3RUWUxpYz0iLCJ3aXRuZXNzZXMiOltdLCJvYmplY3QiOiJsYW8iLCJhY3Rpb24iOiJjcmVhdGUifQ=="),
-      PublicKey(Base64Data("to_klZLtiHV446Fv98OLNdNmi-EP5OaTtbBkotTYLic=")),
-      Signature(Base64Data("2VDJCWg11eNPUvZOnvq5YhqqIKLBcik45n-6o87aUKefmiywagivzD4o_YmjWHzYcb9qg-OgDBZbBNWSUgJICA==")),
-      Hash(Base64Data("f1jTxH8TU2UGUBnikGU3wRTHjhOmIEQVmxZBK55QpsE=")),
-      WitnessSignaturePair(PublicKey(Base64Data("wit1")), Signature(Base64Data("sig1"))) :: WitnessSignaturePair(PublicKey(Base64Data("wit2")), Signature(Base64Data("sig2"))) :: Nil
-    )
 
+    var expected: Message = buildExpected(id, sender, signature, data)
     val message = Message.buildFromJson(source)
 
     message shouldBe a [Message]
     message should equal (expected)
   }
+
+  test("Parser: Empty witness array") {
+     val id: String = "f1jTxH8TU2UGUBnikGU3wRTHjhOmIEQVmxZBK55QpsE="
+    val sender: String ="to_klZLtiHV446Fv98OLNdNmi-EP5OaTtbBkotTYLic="
+    val signature: String = "2VDJCWg11eNPUvZOnvq5YhqqIKLBcik45n-6o87aUKefmiywagivzD4o_YmjWHzYcb9qg-OgDBZbBNWSUgJICA=="
+    val data: String = "eyJjcmVhdGlvbiI6MTYzMTg4NzQ5NiwiaWQiOiJ4aWdzV0ZlUG1veGxkd2txMUt1b0wzT1ZhODl4amdYalRPZEJnSldjR1drPSIsIm5hbWUiOiJoZ2dnZ2dnIiwib3JnYW5pemVyIjoidG9fa2xaTHRpSFY0NDZGdjk4T0xOZE5taS1FUDVPYVR0YkJrb3RUWUxpYz0iLCJ3aXRuZXNzZXMiOltdLCJvYmplY3QiOiJsYW8iLCJhY3Rpb24iOiJjcmVhdGUifQ=="
+    val source: String = s"""{
+                                 |      "message_id": "$id",
+                                 |      "sender": "$sender",
+                                 |      "signature": "$signature",
+                                 |      "data": "$data",
+                                 |      "witness_signatures": []
+                                 |    }""".stripMargin
+
+    val expected: Message = buildExpected(id, sender, signature, data)
+    val message = Message.buildFromJson(source)
+    message shouldBe a [Message]
+    message should equal (expected)
+  }
+
+  test("Parser: Empty source") {
+    val source: String = ""
+    an [spray.json.JsonParser.ParsingException] should be thrownBy(Message.buildFromJson(source))
+  }
+
+  test("Invalid json keys format test"){
+  val set = for { p <- (MsgField.values - WITNESS_SIGNATURES)
+                  t <- Set("[]", "{}", "null", "1")
+                } yield (t,p)
+
+  val setWitness = Set("","1", "{}", "null").map((_,WITNESS_SIGNATURES))
+  forEvery(set union setWitness){
+    case (t: String, p: MsgField) =>  {
+        val source: String = getFormattedString(p,t)
+        an [IllegalArgumentException] should be thrownBy(Message.buildFromJson(source))
+      }
+    }
+  }
```

Comment:
This took me a few moments to understand exactly how it works. It would greatly benefit from a few comments here and there for future students or even future you :)